inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp

```diff
@@ -95,158 +203,327 @@ void MKLDNNPadNode::createPrimitive() {
     }
+
+    prepareParams();
 }
 
-void MKLDNNPadNode::execute(mkldnn::stream strm) {
-    const float *srcData = getParentEdgeAt(0)->getBlob()->cbuffer().as<const float*>() +
-            getParentEdgeAt(0)->getBlob()->getTensorDesc().getBlockingDesc().getOffsetPadding();
-    float* dstData = getChildEdgeAt(0)->getBlob()->buffer().as<float*>() +
-            getChildEdgeAt(0)->getBlob()->getTensorDesc().getBlockingDesc().getOffsetPadding();
+void MKLDNNPadNode::prepareParams() {
+    int beginIdx = 0;
+    int endIdx = padsBegin.size() - 1;
+
+    for (int i = 0; i < padsBegin.size(); ++i) {
+        if (padsBegin[i] != 0 || padsEnd[i] != 0) {
+            beginIdx = i - 1;
+            break;
+        }
+    }
+
+    for (int i = padsBegin.size() - 1; i >= 0; --i) {
+        if (padsBegin[i] != 0 || padsEnd[i] != 0) {
+            endIdx = i;
+            break;
+        }
+    }
+
+    size_t nGluingLastDims = params.dstStrides[std::max(endIdx - 1, 0)];
+    params.nDimsForWork = std::max(endIdx - std::max(beginIdx, 0), 1);
+    params.workAmount = params.dstDims[0];
+    for (int i = 1; i <= beginIdx; ++i) {
+        params.workAmount *= params.dstDims[i];
+        params.dstDims[0] *= params.dstDims[i];
+        params.srcDims[0] *= params.srcDims[i];
+        params.dstStrides[0] /= params.dstDims[i];
+        params.srcStrides[0] /= params.srcDims[i];
+    }
+
+    if (beginIdx > 0) {
+        beginIdx++;
+        params.dstDims.erase(params.dstDims.begin() + 1, params.dstDims.begin() + beginIdx);
+        params.srcDims.erase(params.srcDims.begin() + 1, params.srcDims.begin() + beginIdx);
+        params.dstStrides.erase(params.dstStrides.begin() + 1, params.dstStrides.begin() + beginIdx);
+        params.srcStrides.erase(params.srcStrides.begin() + 1, params.srcStrides.begin() + beginIdx);
+        padsBegin.erase(padsBegin.begin() + 1, padsBegin.begin() + beginIdx);
+        padsEnd.erase(padsEnd.begin() + 1, padsEnd.begin() + beginIdx);
+    }
+    params.workAmount = params.workAmount * params.dstStrides[0] / nGluingLastDims;
+
+    params.lastDstDim = nGluingLastDims;
+    params.shift = params.dstStrides[params.nDimsForWork];
+    if (padMode != CONSTANT || (padMode == CONSTANT && padValue == 0)) {
+        params.lastDstDim *= params.sizeData;
+        params.shift *= params.sizeData;
+    }
 
-    auto srcDims = getParentEdgeAt(0)->getDims().ToSizeVector();
-    auto dstDims = getChildEdgeAt(0)->getDims().ToSizeVector();
+    for (size_t i = 0; i < params.srcDims.size(); ++i)
+        params.srcODims.push_back(padsBegin[i] + params.srcDims[i]);
+
+    if (padMode == REFLECT || padMode == SYMMETRIC) {
+        int shift = padMode == SYMMETRIC ? 1 : 0;
+        for (size_t i = 0; i < params.srcDims.size(); ++i)
+            params.srcDimsForReflectOrSymmetric.push_back(params.srcDims[i] + params.srcODims[i] - 2 + shift);
+    }
+}
 
+void MKLDNNPadNode::execute(mkldnn::stream strm) {
     switch (padMode) {
```

Comment:
I think it would be better to rewrite the switch statement using the Conditional Compilation library.